File: frontend/src/objects/contest/ContestTitle.ts
/** 比赛标题品牌类型；用于列表、详情和编辑表单。 */
export type ContestTitle = string & { readonly __brand: 'ContestTitle' }

/** 解析 helper 的返回结果；成功携带品牌值，失败携带可展示错误，不抛异常。 */
type ParseResult<T> = { ok: true; value: T } | { ok: false; error: string }

/** 创建比赛标题品牌值；调用前必须完成非空和长度校验。 */
function createContestTitle(value: string): ContestTitle {
  /** 注意：这里的 as 只在 parseContestTitle 校验通过后施加品牌类型。 */
  return value as ContestTitle
}

/** 将比赛标题品牌值还原为字符串；无副作用。 */
export function contestTitleValue(title: ContestTitle): string {
  return title
}

/** 解析比赛标题；去除首尾空白并返回结构化校验结果。 */
export function parseContestTitle(rawTitle: string): ParseResult<ContestTitle> {
  const normalized = rawTitle.trim()
  if (!normalized) {
    return { ok: false, error: 'Contest title is required.' }
  }
  if (normalized.length > 120) {
    return { ok: false, error: 'Contest title must be at most 120 characters.' }
  }
  return { ok: true, value: createContestTitle(normalized) }
}

File: backend/src/main/scala/domains/contest/objects/ContestTitle.scala
package domains.contest.objects

import io.circe.{Decoder, Encoder}

/** 比赛标题领域值，保存裁剪后的展示标题。 */
final case class ContestTitle(value: String)

/** 提供比赛标题 JSON codec 和非空/长度校验。 */
object ContestTitle:
  given Encoder[ContestTitle] = Encoder.encodeString.contramap(_.value)
  given Decoder[ContestTitle] = Decoder.decodeString.emap(parse)

  /** 去除首尾空白，要求标题非空且最多 120 字符。 */
  def parse(raw: String): Either[String, ContestTitle] =
    val normalized = raw.trim
    if normalized.isEmpty then Left("Contest title is required.")
    else if normalized.length > 120 then Left("Contest title must be at most 120 characters.")
    else Right(ContestTitle(normalized))
